pkg/controllers/ragengine_controller.go

```diff
@@ -33,14 +54,381 @@ func NewRAGEngineReconciler(client client.Client, scheme *runtime.Scheme, log lo
+	err = c.updateStatusNodeListIfNotMatch(ctx, ragEngineObj, selectedNodes)
+	if err != nil {
+		if updateErr := c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeResourceStatus, metav1.ConditionFalse,
+			"ragengineResourceStatusFailed", err.Error()); updateErr != nil {
+			klog.ErrorS(updateErr, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+			return updateErr
+		}
+		return err
+	}
+
+	if err = c.updateStatusConditionIfNotMatch(ctx, ragEngineObj, kaitov1alpha1.ConditionTypeResourceStatus, metav1.ConditionTrue,
+		"ragengineResourceStatusSuccess", "ragengine resource is ready"); err != nil {
+		klog.ErrorS(err, "failed to update ragengine status", "ragengine", klog.KObj(ragEngineObj))
+		return err
+	}
+
+	return nil
+}
+
+// getAllQualifiedNodes returns all nodes that match the labelSelector and instanceType.
+func (c *RAGEngineReconciler) getAllQualifiedNodes(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) ([]*corev1.Node, error) {
+	var qualifiedNodes []*corev1.Node
+
+	nodeList, err := resources.ListNodes(ctx, c.Client, ragEngineObj.Spec.Compute.LabelSelector.MatchLabels)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(nodeList.Items) == 0 {
+		klog.InfoS("no current nodes match the ragengine resource spec", "ragengine", klog.KObj(ragEngineObj))
+		return nil, nil
+	}
+
+	preferredNodeSet := sets.New(ragEngineObj.Spec.Compute.PreferredNodes...)
+
+	for index := range nodeList.Items {
+		nodeObj := nodeList.Items[index]
+		// skip nodes that are being deleted
+		if nodeObj.DeletionTimestamp != nil {
+			continue
+		}
+
+		// skip nodes that are not ready
+		_, statusRunning := lo.Find(nodeObj.Status.Conditions, func(condition corev1.NodeCondition) bool {
+			return condition.Type == corev1.NodeReady && condition.Status == corev1.ConditionTrue
+		})
+		if !statusRunning {
+			continue
+		}
+
+		// match the preferred node
+		if preferredNodeSet.Has(nodeObj.Name) {
+			qualifiedNodes = append(qualifiedNodes, lo.ToPtr(nodeObj))
+			continue
+		}
+
+		// match the instanceType
+		if nodeObj.Labels[corev1.LabelInstanceTypeStable] == ragEngineObj.Spec.Compute.InstanceType {
+			qualifiedNodes = append(qualifiedNodes, lo.ToPtr(nodeObj))
+		}
+	}
+
+	return qualifiedNodes, nil
+}
+
+// createAndValidateNode creates a new node and validates status.
+func (c *RAGEngineReconciler) createAndValidateNode(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine) (*corev1.Node, error) {
+	var nodeOSDiskSize string
+
+	if nodeOSDiskSize == "" {
+		nodeOSDiskSize = "0" // The default OS size is used
+	}
+
+	if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+		return c.CreateNodeClaim(ctx, ragEngineObj, nodeOSDiskSize)
+	} else {
+		return c.CreateMachine(ctx, ragEngineObj, nodeOSDiskSize)
+	}
+}
+
+func (c *RAGEngineReconciler) CreateMachine(ctx context.Context, ragEngineObj *kaitov1alpha1.RAGEngine, nodeOSDiskSize string) (*corev1.Node, error) {
```

Comment:
same logic to https://github.com/Azure/kaito/blob/2ec45ac3141e955bcf37af78ba5f76f5f7e65cdd/pkg/controllers/workspace_controller.go#L542